### ISSUE TITLE
The "create a pyz of shiv using shiv" example doesn't work

### ISSUE THREAD

**Issue by pfmoore** [linkedin/shiv#69]
As written, the example leaves the shiv pyz file with a shebang pointing at the (presumably temporary) virtual environment used to build it.

The example should probably include a `-p "/usr/bin/env python3"` argument to specify an explicit interpreter.

Actually, given that shiv pyz files are self-contained, there's no real benefit to them ever having shebangs that refer to a virtualenv - maybe it would be worth making the default shebang `/usr/bin/env python3`, or at least locate the "base" Python interpreter for a virtualenv - which can be done as

```python
Path(getattr(sys, 'real_prefix', sys.base_prefix)) / Path(sys.executable).relative_to(sys.prefix)
```

**Comment by sixninetynine** [comment]
whoa! what an awesome trick!
I've definitely kicked myself multiple times for creating a pyz and forgetting to specify `-p`

thanks for the pathlib recipe!

**Comment by sixninetynine** [comment]
@pfmoore so, on windows it appears that the pathlib recipe fails due to `Path(sys.executable).relative_to(sys.prefix)` resulting in "Scripts/python.exe" (what we need is just "python.exe").

I don't know much about doing Python on Windows, is there a safe assumption that can be made about the "Scripts" dir? Like just grabbing the `.stem` attr off the first Path invocation?

**Comment by pfmoore** [comment]
Ah, yes. My mistake - I didn't test it properly.

The problem is that you need to get from the virtualenv/venv back to the original base installation. That's what `sys.base_prefix` gives you (for a venv) and `sys.real_prefix` (for a virtualenv). So all the getattr stuff is covering both bases there.

But they give you the equivalent of `sys.prefix`, *not* `sys.executable`. And the executable gets put in various places - what I was trying to do was avoid hard coding the fact that it's in `Scripts/python.exe` on Windows, and `bin/python` on Unix. But that's for virtual environments, and the whole *point* here is to get a non-virtual location. So I was being dumb. What you actually need is just

```python
Path(getattr(sys, 'real_prefix', sys.base_prefix)) / 'python.exe`
```

That's for Windows - of course, you'll need to miss of the `.exe` suffix on Unix, and maybe do something about `python2` vs `python3` - and maybe Python isn't even in `sys.prefix`, but in `/usr/bin` or `/usr/bin/local`...? Or get everyone to switch to Windows where it's simple :wink:

The more I think about it, the more I'd go for `#!/usr/bin/env python3` (use "python3" or "python2" depending on `sys.version_info`). The Windows launcher handles that, and (hopefully!) it's reliable on Unix.

Or maybe even do what zipapp does, and default to *no* shebang, forcing the user to run the pyz as `python xxx.pyz` or be explicit about what they want via `-p`.

**Comment by sixninetynine** [comment]
Thanks for the extra details!!

My only windows box is a bootcamp partition on a mac mini, but I'll carve out some time to poke at this soon. I definitely want to fix this b/c it's so easy to get bit by and it's really irritating when it happens!

**Comment by sixninetynine** [comment]
@pfmoore what do you think about the reliability of:

```
>>> def _interpreter_path():
...     base_dir = Path(getattr(sys, 'real_prefix', sys.base_prefix))
...     executable = Path(sys.executable).name
...     return next(iter(base_dir.rglob(executable)))
...
>>> _interpreter_path()
PosixPath('/Library/Frameworks/Python.framework/Versions/3.7/bin/python3')
```

Should work just as well on Windows & Linux as it does on OSX, I'd expect, but obviously I'll try it on all three.

**Comment by pfmoore** [comment]
Yeah, that looks really good. Tested on Windows using virtualenv and venv and it seems fine.

**Comment by pfmoore** [comment]
One oddity on Linux. In a Python 3 venv, executables `python` and `python3` are installed. But the base environment only has `python3`. So you get:

```bash
[root@c2874e550a59 /]# python3 -V
Python 3.6.6
[root@c2874e550a59 /]# cat >t.py
from pathlib import Path
import sys

def _interpreter_path():
    base_dir = Path(getattr(sys, 'real_prefix', sys.base_prefix))
    executable = Path(sys.executable).name
    return next(iter(base_dir.rglob(executable)))

i = _interpreter_path()
print(i)
print(i.exists())
[root@c2874e550a59 /]# python3 t.py
/usr/bin/python3
True
[root@c2874e550a59 /]# python3 -m venv x
x/[root@c2874e550a59 /]# x/bin/python t.py
Traceback (most recent call last):
  File "t.py", line 9, in <module>
    i = _interpreter_path()
  File "t.py", line 7, in _interpreter_path
    return next(iter(base_dir.rglob(executable)))
StopIteration
[root@c2874e550a59 /]# x/bin/python3 t.py
/usr/bin/python3
True
```

Probably not a big deal, although *maybe* you could look for `next(iter(base_dir.rglob(executable + str(sys.version_info[0]))))` if the initial check finds nothing. At some point, though, you have to stop trying  and tell the user to just use `-p`...

**Comment by sixninetynine** [comment]
Good catch!

> At some point, though, you have to stop trying and tell the user to just use -p...

I agree with this sentiment wholeheartedly, haha, however this could be a good middle ground:

```
def _interpreter_path(append_version: bool=False) -> str:

    base_dir = Path(getattr(sys, "real_prefix", sys.base_prefix)).resolve()
    name = Path(sys.executable).name

    if append_version:
        name += str(sys.version_info.major)

    try:
        return str(next(iter(base_dir.rglob(name))))
    except StopIteration:
        if not append_version:
            return _interpreter_path(append_version=True)
        raise
```

**Comment by pfmoore** [comment]
I'd still be inclined to add a final, final fallback of `sys.executable` (possibly with a warning), just because it's silly to fail as a result of trying too hard when there's an obvious option to choose.

**Comment by sixninetynine** [comment]
instead of the `raise`, yeah I agree. And that's the current behavior anyway! 

cool, will probably push #70 with that incorporated later today or tomorrow (or forever hold your peace! 😆  )

**Comment by sixninetynine** [comment]
released on pypi @ 0.0.37
